// @simple-release/core/dist/project/monorepo.js
import semver from 'semver';
import { packagePrefix } from '@conventional-changelog/git-client';
import { bumpDefaultOptions, Project } from './project.js';
export * from './monorepo.types.js';
export class MonorepoProject extends Project {
    /**
     * The mode of the monorepo.
     * If mode is 'fixed', all projects will be bumped to the same version.
     * If mode is 'independent', each project will be bumped to its own version.
     */
    mode;
    projectsMutex;
    /**
     * Creates a new instance of the monorepo project.
     * @param options - The options to use for the monorepo project.
     */
    constructor(options) {
        const { mode, compose } = options;
        super({
            ...options,
            compose: compose
                ? (manifest, isRoot = true) => compose(manifest, isRoot)
                : manifest => manifest
        });
        this.mode = mode;
    }
    /**
     * Get the projects in the monorepo.
     * @yields The projects in the monorepo.
     */
    async *getProjects() {
        if (!this.projectsMutex) {
            const { compose, getProjects } = this.options;
            const options = {
                compose: (manifest) => compose(manifest, false),
                gitClient: this.gitClient,
                manifest: this.manifest
            };
            const projects = [];
            let resolve;
            this.projectsMutex = new Promise((r) => {
                resolve = r;
            });
            for await (const project of getProjects(options)) {
                const isPrivate = await project.manifest.isPrivate();
                if (!isPrivate) {
                    projects.push(project);
                    yield project;
                }
            }
            resolve(projects);
            return;
        }
        yield* await this.projectsMutex;
    }
    /**
     * Get scope from project name.
     * @param name - The name of the project.
     * @returns The scope of the project.
     */
    async getScope(name) {
        if (this.options.scope) {
            return await this.options.scope(name, await this.manifest.getName());
        }
        return name.replace(/@[^/]+\//, '');
    }
    /**
     * Get tag prefix from scope.
     * @param scope - The scope of the project.
     * @returns The tag prefix for the project.
     */
    async getTagPrefix(scope) {
        if (this.options.tagPrefix) {
            return await this.options.tagPrefix(scope);
        }
        if (this.mode === 'fixed') {
            return '';
        }
        return packagePrefix(scope);
    }
    getIndependentCommitMessage() {
        const { versionUpdates } = this;
        if (versionUpdates.length === 0) {
            throw new Error('Nothing to commit');
        }
        return `chore(release): monorepo release\n\n${versionUpdates
            .map(({ name, to }) => `- ${name}@${to}`)
            .join('\n')}`;
    }
    getCommitMessage() {
        const { mode } = this;
        if (mode === 'fixed') {
            return super.getCommitMessage();
        }
        return this.getIndependentCommitMessage();
    }
    async getIndependentTags() {
        const tags = [];
        for await (const project of this.getProjects()) {
            const { manifest } = project;
            const name = await manifest.getName();
            const scope = await this.getScope(name);
            const tagPrefix = await this.getTagPrefix(scope);
            tags.push(...await project.getTags({
                tagPrefix
            }));
        }
        return tags;
    }
    async getTags() {
        const { mode } = this;
        if (mode === 'fixed') {
            return super.getTags();
        }
        return this.getIndependentTags();
    }
    async getIndependentReleaseData() {
        const data = [];
        for await (const project of this.getProjects()) {
            const { manifest } = project;
            const name = await manifest.getName();
            const scope = await this.getScope(name);
            const tagPrefix = await this.getTagPrefix(scope);
            const releaseData = await project.getReleaseData({
                tagPrefix
            });
            data.push(...releaseData.map(data => ({
                ...data,
                title: `${name}: ${data.title}`
            })));
        }
        return data;
    }
    async getReleaseData() {
        const { mode } = this;
        if (mode === 'fixed') {
            return super.getReleaseData();
        }
        return this.getIndependentReleaseData();
    }
    async getBumpOptions(project, options, baseVersion) {
        const { preset = bumpDefaultOptions.preset, extraScopes = [], byProject, ...bumpOptions } = options;
        const { manifest } = project;
        const name = await manifest.getName();
        const scope = await this.getScope(name);
        const tagPrefix = await this.getTagPrefix(scope);
        const projectPreset = {
            ...typeof preset === 'string'
                ? {
                    name: preset
                }
                : preset,
            scope: [scope, ...extraScopes]
        };
        const projectBumpOptions = {
            ...bumpOptions,
            ...byProject?.[name],
            preset: projectPreset,
            tagPrefix,
            baseVersion
        };
        return projectBumpOptions;
    }
    getNextVersion(options = {}) {
        if (this.mode === 'fixed' && options.version) {
            return super.getNextVersion(options);
        }
        throw new Error('Monorepo project does not support getNextVersion');
    }
    async independentBump(options = {}) {
        let hasBump = false;
        for await (const project of this.getProjects()) {
            const projectBumpOptions = await this.getBumpOptions(project, options);
            hasBump = await project.bump(projectBumpOptions) || hasBump;
            this.changedFiles.push(...project.changedFiles);
            this.versionUpdates.push(...project.versionUpdates);
        }
        return hasBump;
    }
    async fixedBump(options = {}) {
        const { force } = options;
        const updatedProjects = [];
        const baseVersion = await this.manifest.getVersion();
        let hasBump = false;
        let fixedVersion;
        for await (const project of this.getProjects()) {
            const projectBumpOptions = await this.getBumpOptions(project, options, baseVersion);
            const version = await project.getNextVersion(projectBumpOptions);
            if (version || force) {
                updatedProjects.push({
                    project,
                    options: projectBumpOptions
                });
                if (version && (!fixedVersion || semver.gt(version, fixedVersion))) {
                    fixedVersion = version;
                }
            }
        }
        if (fixedVersion) {
            hasBump = await super.bump({
                ...options,
                forcePrivate: true,
                version: fixedVersion,
                tagPrefix: await this.getTagPrefix('')
            });
            for (const { project, options } of updatedProjects) {
                hasBump = await project.bump({
                    ...options,
                    version: fixedVersion
                }) || hasBump;
                this.changedFiles.push(...project.changedFiles);
                this.versionUpdates.push(...project.versionUpdates);
            }
        }
        return hasBump;
    }
    async bump(options = {}) {
        const { mode } = this;
        if (mode === 'fixed') {
            return this.fixedBump(options);
        }
        return this.independentBump(options);
    }
}
