// @simple-release/core/dist/logger.js
/* eslint-disable max-classes-per-file */
function defaultPrinter({ dryRun, stage, message }) {
    // eslint-disable-next-line no-console
    console.log(`${dryRun ? `[dry-run]` : ''}[${stage}]:`, message);
}
/**
 * A child messages logger that bounds to a specific stage.
 */
export class ChildLogger {
    parent;
    stage;
    /**
     * Creates a child logger instance.
     * @param parent - The parent logger to inherit options from.
     * @param stage - The stage of the process this logger is associated with.
     */
    constructor(parent, stage) {
        this.parent = parent;
        this.stage = stage;
    }
    /**
     * Logs an info message.
     * @param message - The message to log.
     */
    info(message) {
        this.parent.info(this.stage, message);
    }
    /**
     * Logs a verbose message.
     * @param message - The message to log.
     */
    verbose(message) {
        this.parent.verbose(this.stage, message);
    }
}
/**
 * A messages logger.
 */
export class Logger {
    options;
    printer;
    /**
     * Creates a logger instance.
     * @param options - The logger options.
     */
    constructor(options = {}) {
        this.options = options;
        const { printer = defaultPrinter } = options;
        this.printer = printer;
    }
    /**
     * Logs an info message.
     * @param stage - The stage of the process.
     * @param message - The message to log.
     */
    info(stage, message) {
        const { silent, dryRun } = this.options;
        if (!silent) {
            this.printer({
                dryRun,
                stage,
                message
            });
        }
    }
    /**
     * Logs a verbose message.
     * @param stage - The stage of the process.
     * @param message - The message to log.
     */
    verbose(stage, message) {
        const { silent, verbose, dryRun } = this.options;
        if (!silent && (verbose || dryRun)) {
            this.printer({
                dryRun,
                stage,
                message
            });
        }
    }
    /**
     * Creates a child logger for a specific stage.
     * @param stage - The stage of the process.
     * @returns A child logger for the specified stage.
     */
    createChild(stage) {
        return new ChildLogger(this, stage);
    }
}
