// @simple-release/config/dist/index.js
import fs from 'fs/promises';
import { findUp } from 'find-up-simple';
const VARIANTS = [
    '.simple-release.js',
    '.simple-release.mjs',
    '.simple-release.cjs',
    '.simple-release.json'
];
function validate(target, rules) {
    for (const [rule, required] of Object.entries(rules)) {
        if (required && target[rule] === undefined) {
            throw new Error(`Faild to load config: '${rule}' is required`);
        }
    }
}
const IMPORT_QUERY_REGEX = /^(@?[^@#]*)(@[^#]+)?(#.*)?$/;
export function parseImportQuery(importPath) {
    const match = importPath.match(IMPORT_QUERY_REGEX);
    if (!match) {
        return {
            path: importPath
        };
    }
    const [, path, version, symbol] = match;
    return {
        path,
        version: version?.slice(1),
        symbol: symbol?.slice(1)
    };
}
export async function loadClass(queryWithOptions, config, loader = _ => import(_)) {
    const [query, options] = Array.isArray(queryWithOptions)
        ? queryWithOptions
        : [queryWithOptions];
    const { path, version, symbol } = parseImportQuery(query);
    const module = await loader(path, version, config);
    const inst = new module[symbol || 'default'](options);
    return inst;
}
export function isQuery(value) {
    return typeof value === 'string' || (Array.isArray(value)
        && value.length === 2
        && typeof value[0] === 'string'
        && value[1] && typeof value[1] === 'object');
}
export function getQuery(queryWithOptions) {
    return isQuery(queryWithOptions)
        ? typeof queryWithOptions === 'string'
            ? queryWithOptions
            : queryWithOptions[0]
        : null;
}
async function loadAndSetIfQuery(config, key, loader) {
    const value = config[key];
    if (value && isQuery(value)) {
        // eslint-disable-next-line require-atomic-updates
        config[key] = await loadClass(value, config, loader);
    }
}
async function importConfig(path) {
    if (path.endsWith('.json')) {
        const json = await fs.readFile(path, 'utf-8');
        const config = JSON.parse(json);
        return config;
    }
    const module = await import(path);
    const config = 'default' in module ? module.default : module;
    return config;
}
export async function load(requirements = {}, loader) {
    const { config: configRequired, ...reqs } = requirements;
    for (const variant of VARIANTS) {
        const foundPath = await findUp(variant);
        if (foundPath) {
            try {
                const config = await importConfig(foundPath);
                validate(config, reqs);
                await loadAndSetIfQuery(config, 'project', loader);
                await loadAndSetIfQuery(config, 'hosting', loader);
                return config;
            }
            catch (err) {
                if (configRequired) {
                    throw err;
                }
                return null;
            }
        }
    }
    if (configRequired) {
        throw new Error('Config not found');
    }
    return null;
}
